src/rgw/rgw_gc.cc

```diff
@@ -65,7 +65,64 @@ int RGWGC::tag_index(const string& tag)
   return rgw_shards_mod(XXH64(tag.c_str(), tag.size(), seed), max_objs);
 }
 
-int RGWGC::send_chain(cls_rgw_obj_chain& chain, const string& tag)
+int RGWGC::send_split_chain(const cls_rgw_obj_chain& chain, const std::string& tag)
+{
+  constexpr int FACTOR_INCREASE = 3; //approximate value by which total size of chain, tag and time increases after encoding into bufferlist
+  ldpp_dout(this, 20) << "RGWGC::send_split_chain - tag is: " << tag << dendl;
+
+  if (cct->_conf->osd_max_write_size) {
+    cls_rgw_obj_chain broken_chain;
+    bool chain_sent = false;
+    long unsigned int tag_size = tag.length(), expiration_time_size = sizeof(ceph::real_time);
+    long unsigned int adjusted_osd_max_write_size = ((cct->_conf->osd_max_write_size << 20)/FACTOR_INCREASE); //approx upper limit
+    ldpp_dout(this, 20) << "RGWGC::send_split_chain - adjusted_osd_max_write_size is: " << adjusted_osd_max_write_size << dendl;
+    long unsigned int cumulative_size = 0;
+    for (auto it = chain.objs.begin(); it != chain.objs.end(); it++) {
+      long unsigned int size = it->pool.size() + it->loc.size() + it->key.name.size() + it->key.instance.size();
+      cumulative_size += size;
+      long unsigned int total_size = cumulative_size + tag_size + expiration_time_size;
+      ldpp_dout(this, 20) << "RGWGC::send_split_chain - cumulative_size is: " << cumulative_size << dendl;
+      ldpp_dout(this, 20) << "RGWGC::send_split_chain - total_size is: " << total_size << dendl;
+
+      if (total_size < adjusted_osd_max_write_size) { //continue adding to chain
+        broken_chain.objs.emplace_back(*it);
+        chain_sent = false;
+        ldpp_dout(this, 20) << "RGWGC::send_split_chain - less than, add to broken chain and continue" << dendl;
+        continue;
+      }
+      if (total_size == adjusted_osd_max_write_size) { //add to chain and send to gc
```

Comment:
I don't imagine we hit the exact size very often...

Reply:
meaning the actual send usually happens in the overshoot case.  The backtracking step looks inexpensive, so maybe that's fine.